utils/run-clang-tidy.py

```diff
@@ -122,33 +131,302 @@ def make_absolute(f, directory):
+            os.remove(self.version_file)
+        if os.path.isfile(self.config_dump_file):
+            os.remove(self.config_dump_file)
+        if os.path.isfile(self.checks_file):
+            os.remove(self.checks_file)
+        if os.path.isfile(self.config_hash_file):
+            os.remove(self.config_hash_file)
+        if os.path.isfile(self.passed_file):
+            os.remove(self.passed_file)
+        with os.scandir(self.failed_dir) as dir:
+            for entry in dir:
+                if entry.is_file() and self.failed_files_regexp.match(entry.name):
+                    os.remove(entry)
+        self.passed = dict()
+        self.failed = set()
+
+    def lookup(self, hash):
+        """Look up hash in the cache.
+
+        Meaning of return values:
+
+        found (bool), passed (bool), stdout (string), stderr (string)
+        """
+        if hash in self.passed:
+            self.hits += 1
+            self.passed[hash] = int(time.time())
+            return (True, True, '', 'Cache hit as OK\n')
+        if hash in self.failed:
+            self.hits += 1
+            out = None
+            err = None
+            out_fn = os.path.join(self.failed_dir, hash + '.stdout')
+            err_fn = os.path.join(self.failed_dir, hash + '.stderr')
+            try:
+                if os.path.isfile(out_fn):
+                    with open(out_fn, 'r', encoding = 'utf-8') as of:
+                        out = of.read()
+                    os.utime(out_fn)
+                if os.path.isfile(err_fn):
+                    with open(err_fn, 'r', encoding = 'utf-8') as ef:
+                        err = ef.read()
+                    os.utime(err_fn)
+            except OSError:
+                pass
+            if out and err:
+                return (True, False, out, 'Retrieved from the cache.\n' + err)
+            # This shouldn't happen, but if it does, there's no harm in making
+            # it a cache miss, so fall through.
+        self.misses += 1
+        return (False, False, None, None)
+
+    def add_passed(self, hash):
+        self.passed[hash] = int(time.time())
+
+    def add_failed(self, hash, stdout, stderr):
+        self.failed.add(hash)
+        out_fn = os.path.join(self.failed_dir, hash + '.stdout')
+        err_fn = os.path.join(self.failed_dir, hash + '.stderr')
+        try:
+            with open(out_fn, 'w', encoding = 'utf-8') as of:
+                of.write(stdout)
+            with open(err_fn, 'w', encoding = 'utf-8') as ef:
+                ef.write(stderr)
+            return True
+        except OSError as error:
+            print(error, file=sys.stderr)
+            print(f'WARNING: Could not store outputs in cache for {hash}\n',
+                  file=sys.stderr)
+            return False
+
+
+def get_hash(text_to_hash):
+    return hashlib.sha256(text_to_hash.encode('utf-8')).hexdigest()
+
+
+extra_before = []
+extra_after = []
+
+def pp_hash(file, command, dir):
```

Comment:
Can you add a short comment, what this function does? It seems to be the core of the cache key generation.